from primer_explorer.config import SHORT_PRODUCTS_CUTOFF


def generate_pair_stats(primer_pair, count, short_products_cutoff=SHORT_PRODUCTS_CUTOFF):
    report = []
    report.append("{},{}".format(primer_pair[0].decode(), primer_pair[1].decode()))
    report.append("-" * 20)
    report.append("Total Number of products:\t{}".format(str(count["total_products"])))
    viable_products_ratio = float(count["viable_products"] / count["total_products"])
    report.append("Number of viable pcr products:\t{}\t{}".format(str(count["viable_products"]), str(viable_products_ratio)))
    number_of_no_viable_products_removed = count["total_products"] - count["viable_products"]
    report.append("Number of short products:\t{}\t{:.2f}".format(str(number_of_no_viable_products_removed),
                                                                 number_of_no_viable_products_removed / count["total_products"]))
    report.append("Number of euchromatic effective products:\t{}".format(str(count["euchromatin_products"])))
    report.append("Number of effective euchromatic nucleotides:\t{}".format(str(count["euchromatin_nucleotides"])))
    report.append("Number of heterochromatic effective products:\t{}".format(str(count["heterochromatin_products"])))
    report.append("Number of effective heterochromatic nucleotides:\t{}".format(str(count["heterochromatin_nucleotides"])))
    report.append("Number of mixed effective products:\t{}".format(str(count["mixed_products"])))
    report.append("Number of mixed effective nucleotides:\t{}".format(str(count["mixed_nucleotides"])))
    report.append("Ratio euchromatin (without mixed):\t{:.2f}".format(count["euchromatin_products"] / (count["viable_products"] - count["mixed_products"])))
    report.append("Ratio heterochromatin (without mixed):\t{:.2f}".format(count["heterochromatin_products"] / (count["viable_products"] - count["mixed_products"])))
    if viable_products_ratio <= short_products_cutoff:
        report.append("WARNING: too much products generated by this pair are no effective")
    return report


def write_detailed_report(report_fhand, stats):
    report = []
    for idx, set_stats in stats.items():
        primers = set_stats['primers']
        report.append("PRIMER SET {}".format(str(idx)))
        report.append('Primers: ' + ', '.join([p.decode() for p in primers]))
        report.append("#" * 30)
        for pair, counts in set_stats['stats'].items():
            report.extend(generate_pair_stats(pair, counts))
            report.append("-" * 20)

    report_fhand.write("\n".join(report))
    report_fhand.flush()


def write_gff_report(gff_results, output_fhand):
    report = []
    report.append("PRIMER PAIR\tEXONS\tINTRONS\tNUM_PCR_PRODUCTS")
    for primer_pair, results in gff_results.items():
        report.append("{}\t{}\t{}\t{}".format(primer_pair, results["exon"],
                                              results["intron"],
                                              results["num_pcr_products"]))
    output_fhand.write("\n".join(report).encode())
    output_fhand.flush()
    output_fhand.close()
